File: src/main/kotlin/customfit/ai/kotlinclient/platform/BackgroundStateMonitor.kt
package customfit.ai.kotlinclient.platform

/** Represents the application's state */
enum class AppState {
    /** App is in the foreground and visible to the user */
    FOREGROUND,

    /** App is in the background (not visible) */
    BACKGROUND
}

/** Interface for battery state */
data class BatteryState(
        val isLow: Boolean,
        val isCharging: Boolean,
        val level: Float // 0.0 to 1.0
)

/** Interface for objects that will be notified when app state changes */
interface AppStateListener {
    /**
     * Called when the app state changes
     *
     * @param state the new app state
     */
    fun onAppStateChange(state: AppState)
}

/** Interface for objects that will be notified when battery state changes */
interface BatteryStateListener {
    /**
     * Called when the battery state changes
     *
     * @param state the new battery state
     */
    fun onBatteryStateChange(state: BatteryState)
}

/**
 * Interface for monitoring application foreground/background state and battery level.
 * Implementation will be platform-specific.
 */
interface BackgroundStateMonitor {
    /** Get the current app state */
    fun getCurrentAppState(): AppState

    /** Get the current battery state */
    fun getCurrentBatteryState(): BatteryState

    /** Register a listener for app state changes */
    fun addAppStateListener(listener: AppStateListener)

    /** Unregister a listener for app state changes */
    fun removeAppStateListener(listener: AppStateListener)

    /** Register a listener for battery state changes */
    fun addBatteryStateListener(listener: BatteryStateListener)

    /** Unregister a listener for battery state changes */
    fun removeBatteryStateListener(listener: BatteryStateListener)

    /** Clean up resources */
    fun shutdown()
}

File: customfit-flutter-client-sdk/lib/src/platform/background_state_monitor.dart
import 'app_state.dart';
import 'app_state_listener.dart';
import 'battery_state.dart';
import 'battery_state_listener.dart';

/// Abstract class for background state monitoring.
abstract class BackgroundStateMonitor {
  /// Add app state listener
  void addAppStateListener(AppStateListener listener);

  /// Remove app state listener
  void removeAppStateListener(AppStateListener listener);

  /// Add battery state listener
  void addBatteryStateListener(BatteryStateListener listener);

  /// Remove battery state listener
  void removeBatteryStateListener(BatteryStateListener listener);

  /// Get current app state
  AppState getCurrentAppState();

  /// Get current battery state
  BatteryState getCurrentBatteryState();

  /// Get current battery level (0-100)
  int getCurrentBatteryLevel();

  /// Clean up resources
  void shutdown();
}
